crates/bevy_pbr/src/meshlet/cull_meshlets.wgsl

```diff
@@ -48,61 +73,59 @@ fn cull_meshlets(@builtin(global_invocation_id) cluster_id: vec3<u32>) {
+        meshlet_visible = sphere_depth >= occluder_depth;
+    } else {
+        // Perspective
+        let sphere_depth = -view.projection[3][2] / (culling_bounding_sphere_center_view_space.z + culling_bounding_sphere_radius);
+        meshlet_visible = sphere_depth >= occluder_depth;
     }
 
-#ifdef MESHLET_SECOND_CULLING_PASS
-    // In the second culling pass, cull against the depth pyramid generated from the first pass
+    // Write if the cluster should be occlusion tested in the second pass
+#ifdef MESHLET_FIRST_CULLING_PASS
+    let second_pass_candidate = u32(!meshlet_visible) << cluster_id % 32u;
+    atomicOr(&meshlet_second_pass_candidates[cluster_id / 32u], second_pass_candidate);
+#endif
+
+    // Append a list of this cluster's triangles to draw if not culled
     if meshlet_visible {
-        let culling_bounding_sphere_center_view_space = (view.inverse_view * vec4(culling_bounding_sphere_center.xyz, 1.0)).xyz;
-        let aabb = project_view_space_sphere_to_screen_space_aabb(culling_bounding_sphere_center_view_space, culling_bounding_sphere_radius);
-
-        // Halve the AABB size because the first depth mip resampling pass cut the full screen resolution into a power of two conservatively
-        let depth_pyramid_size_mip_0 = vec2<f32>(textureDimensions(depth_pyramid, 0)) * 0.5;
-        let width = (aabb.z - aabb.x) * depth_pyramid_size_mip_0.x;
-        let height = (aabb.w - aabb.y) * depth_pyramid_size_mip_0.y;
-        let depth_level = max(0, i32(ceil(log2(max(width, height))))); // TODO: Naga doesn't like this being a u32
-        let depth_pyramid_size = vec2<f32>(textureDimensions(depth_pyramid, depth_level));
-        let aabb_top_left = vec2<u32>(aabb.xy * depth_pyramid_size);
-
-        let depth_quad_a = textureLoad(depth_pyramid, aabb_top_left, depth_level).x;
-        let depth_quad_b = textureLoad(depth_pyramid, aabb_top_left + vec2(1u, 0u), depth_level).x;
-        let depth_quad_c = textureLoad(depth_pyramid, aabb_top_left + vec2(0u, 1u), depth_level).x;
-        let depth_quad_d = textureLoad(depth_pyramid, aabb_top_left + vec2(1u, 1u), depth_level).x;
-
-        let occluder_depth = min(min(depth_quad_a, depth_quad_b), min(depth_quad_c, depth_quad_d));
-        if view.projection[3][3] == 1.0 {
-            // Orthographic
-            let sphere_depth = view.projection[3][2] + (culling_bounding_sphere_center_view_space.z + culling_bounding_sphere_radius) * view.projection[2][2];
-            meshlet_visible &= sphere_depth >= occluder_depth;
-        } else {
-            // Perspective
-            let sphere_depth = -view.projection[3][2] / (culling_bounding_sphere_center_view_space.z + culling_bounding_sphere_radius);
-            meshlet_visible &= sphere_depth >= occluder_depth;
+        let meshlet_triangle_count = meshlets[meshlet_id].triangle_count;
+        let buffer_start = atomicAdd(&draw_indirect_args.vertex_count, meshlet_triangle_count * 3u) / 3u;
```

Comment:
is it not possible to skip the muls and divs by 3 here and just multiply afterwards at the last possible moment?

Reply:
We'd need an extra pass to just multiply by 3, I doubt it's worth it just to save like 2 instructions.